posthog/models/historical_version.py

```diff
@@ -0,0 +1,108 @@
+import datetime
+import json
+
+from django.db import models
+from django.utils import timezone
+
+from posthog.models import FeatureFlag
+from posthog.models.utils import UUIDModel
+
+
+class HistoricalVersionJSONEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, models.base.ModelState):
+            return None
+        if isinstance(obj, FeatureFlag):
+            return obj.__dict__
+        if isinstance(obj, datetime.datetime):
+            return obj.isoformat()
+
+        return json.JSONEncoder.default(self, obj)
+
+
+class HistoricalVersion(UUIDModel):
```

Comment:
non-blocking

I should also mention, I'm a bit apprehensive about making this table shared across all types. That might be desired for an audit log type query pattern but history for a single object as far as I can see doesn't need a shared table.

Instead, we can have a shared sub-structure on which e.g. `load_history` can operate, whilst keeping data separate.

Reply:
I'm expecting queries like "changes to insights" but also "changes by user" or "last N changes" e.g. https://www.figma.com/file/gQBj9YnNgD8YW4nBwCVLZf/PostHog-App?node-id=7436%3A40010

What's the source of the apprehension?

Reply:
The apprehension comes from having the schema of `state` owned by different parts of the application, and this table potentially being a hotspot/single point of failure.

I don't have anything more solid than that, I don't want to hold up proceedings.